File: src/DotNet/SoftMemes.Napl/Expressions.fs
namespace SoftMemes.Napl

open System

type NaplValue
    = BooleanValue of bool
    | IntegerValue of int
    | FloatValue of float
    | StringValue of string

[<CustomEquality>]
[<CustomComparison>]
type NaplParameter = { Type : NaplType; Name : string }
    with
        override param.Equals(other) = obj.ReferenceEquals(param, other)
        interface IComparable with
            member param.CompareTo(otherParam) =
                compare (param.GetHashCode()) (otherParam.GetHashCode())

[<StructuralEquality>]
[<StructuralComparison>]
type NaplExpression<'t> = { Annotation : 't; Expression : NaplExpression'<'t> }
and NaplExpression'<'t>
    = ValueExpression of NaplValue
    | OperatorExpression of NaplOperator * NaplExpression<'t> list
    | InstantiateExpression of NaplType * NaplExpression<'t> list
    | LambdaExpression of NaplParameter list * NaplExpression<'t>
    | ParameterExpression of NaplParameter
    | ApplyExpression of NaplExpression<'t> * NaplExpression<'t> list

File: src/Java/src/com/softmemes/napl/Expressions.scala
package com.softmemes.napl

abstract class NaplValue
case class BooleanValue(value : Boolean) extends NaplValue
case class IntegerValue(value : Int) extends NaplValue
case class FloatValue(value : Double) extends NaplValue
case class StringValue(value : String) extends NaplValue

// TODO: Reference equality / hashing?
class NaplParameter(naplType : NaplType, name : String)

case class NaplExpression[T](tag : T, expression : NaplExpression_[T]) extends NaplExpression_[T]

abstract class NaplExpression_[T]
case class LambdaExpression[T](params : Vector[NaplParameter], body : NaplExpression[T]) extends NaplExpression_[T]
case class ValueExpression[T](value : NaplValue) extends NaplExpression_[T]
case class ParameterExpression[T](param : NaplParameter) extends NaplExpression_[T]
case class OperatorExpression[T](opr : NaplOperator, args : Vector[NaplExpression[T]]) extends NaplExpression_[T]
case class ApplyExpression[T](funcExpr : NaplExpression[T], argExprs : Vector[NaplExpression[T]]) extends NaplExpression_[T]
case class InstantiateExpression[T](t : NaplType, exprs :  Vector[NaplExpression[T]]) extends NaplExpression_[T]
